pkg/scheduler/service/bookkeeper_test.go

```diff
@@ -85,3 +92,116 @@ func TestBookkeeper(t *testing.T) {
 		require.NoError(t, reconRepo.RemoveReconciliation(recon.SchedulingID))
 	}
 }
+
+func TestBookkeeperParallel(t *testing.T) {
+	tests := []struct {
+		name        string
+		markOpsDone bool
+		customFunc  string
+		errMessage  string
+		errCount    int
+	}{
+		{name: "Mark two operations as orphan in multiple parallel threads", markOpsDone: false, customFunc: "markOrphanOperations", errMessage: "Bookkeeper failed to update status of orphan operation", errCount: 98},
+		{name: "Finish two operations in multiple parallel threads", markOpsDone: true, customFunc: "finishReconciliation", errMessage: "Bookkeeper failed to update cluster", errCount: 49},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			//make sure persistent db is empty
+			err := os.RemoveAll("./test/")
+			require.NoError(t, err)
+			//initialize WaitGroup
+			var wg sync.WaitGroup
+
+			//create mock database connection
+			dbConn := db.NewTestConnection(t)
+			//prepare inventory
+			inventory, err := cluster.NewInventory(dbConn, true, cluster.MetricsCollectorMock{})
+			require.NoError(t, err)
+
+			//add cluster to inventory
+			clusterState, err := inventory.CreateOrUpdate(1, &keb.Cluster{
+				Kubeconfig: "123",
+				KymaConfig: keb.KymaConfig{
+					Components: []keb.Component{
+						{
+							Component:     "dummy",
+							Configuration: nil,
+							Namespace:     "kyma-system",
+						},
+					},
+					Profile: "",
+					Version: "1.2.3",
+				},
+				RuntimeID: uuid.NewString(),
+			})
+			require.NoError(t, err)
+
+			//trigger reconciliation for cluster
+			reconRepo, err := reconciliation.NewPersistedReconciliationRepository(dbConn, true)
+			require.NoError(t, err)
+			reconEntity, err := reconRepo.CreateReconciliation(clusterState, nil)
+			require.NoError(t, err)
+			require.NotEmpty(t, reconEntity.Lock)
+			require.False(t, reconEntity.Finished)
+
+			if tc.markOpsDone {
+				//mark all operations to be done
+				opEntities, err := reconRepo.GetOperations(reconEntity.SchedulingID)
+				require.NoError(t, err)
+				for _, opEntity := range opEntities {
+					err := reconRepo.UpdateOperationState(opEntity.SchedulingID, opEntity.CorrelationID, model.OperationStateDone)
+					require.NoError(t, err)
+				}
+			}
+
+			//initialize logger including error channel
+			errChannel := make(chan error, 100)
+			bookkeeperLogger := zaptest.NewLogger(t, zaptest.WrapOptions(zap.Hooks(func(e zapcore.Entry) error {
```

Comment:
General a good idea to inject some interception logic into another object - but failure prone if someone changes latter the log behaviour.

If a function has to be tested but it's not possible to test it, it indicates a bad testable code base which requires a refactoring.

Would recommend following:

* the Bookkeeper has already too many responsibilities bundled in one place: let's introduce a responsibility pattern and extract the Orphan-Ops detection and Finish-reconciliation logic dedicated chain-steps.
* Each step can be implemented in way that it becomes testable.

Reply:
Done